torch/distributed/optim/named_optimizer.py

```diff
@@ -0,0 +1,244 @@
+import logging
+import warnings
+
+from copy import deepcopy
+from typing import Any, Collection, Dict, List, Mapping, Union
+
+import torch
+from torch import optim
+from torch.distributed._shard.sharded_tensor import ShardedTensor
+
+
+__all__ = ["NamedOptimizer"]
+
+logger = logging.getLogger(__name__)
+
+
+class NamedOptimizer(optim.Optimizer):
+    """
+    NamedOptimizer takes a dict of parameters and exposes state_dict by parameter key.
+    We replace the original key (number) in an optim to the FQN string. User can
+    initialize the optim as they initialize a PyTorch optim, the only difference is
+    that they also need to passed in the FQN of each parameters.
+
+    Args:
+        named_parameters (Mapping[str, Union[torch.Tensor, ShardedTensor]]):
+            Parameters of the module and its FQN.
+        optimizer_class (optim.Optimizer):
+            the class of optimizer to instantiate.
+        param_groups (Collection[Mapping[str, Any]]):
+            param_groups to pass to optimizer if specified.
+            Default: None
+        args: arguments to pass to the optimizer constructor.
+        kwargs: arguments to pass to the optimizer constructor.
+
+    Example::
+        >>> # xdoctest: +SKIP("distributed")
+        >>> from torch import optim
+        >>> from torch.distributed.optim import NamedOptimizer
+        >>>
+        >>> # Define the named optimizer.
+        >>> m = Model(...)
+        >>> named_optim = NamedOptimizer(m.named_parameters(), optim.SGD)
+        >>> # Forward pass + backward pass.
+        >>> named_optim.step()
+        >>> ...
+        >>> # Call state_dict for the named optimizer returns a FQN state_dict.
+        >>> named_optim.state_dict()
+
+    __ TODO: Add tutorial for NamedOptimizer.
+    """
+
+    def __init__(
+        self,
+        named_parameters: Mapping[str, Union[torch.Tensor, ShardedTensor]],
+        optimizer_class: optim.Optimizer,
+        param_groups: Collection[Mapping[str, Any]] = None,
+        *args,
+        **kwargs,
+    ) -> None:
+        torch._C._log_api_usage_once("torch.distributed.optim.NamedOptimizer")
+        self.param_groups: Collection[Mapping[str, Any]] = param_groups  # type: ignore[assignment]
+        self.named_parameters = dict(named_parameters)
+        params_optimizer = (
+            self.named_parameters.values() if param_groups is None else param_groups
+        )
+        self._optimizer = optimizer_class(  # type: ignore[operator]
+            params_optimizer,
+            *args,
+            **kwargs,
+        )
+        if param_groups is None:
+            self.param_keys_order = list(self.named_parameters.keys())
```

Comment:
Hmm. I am not sure about the naming here (`param_keys_order`). This seems like the ordered parameter keys, which are the FQNs right?

Maybe we rename this to `ordered_param_names` or `ordered_param_keys`? Does this need to be public?